### ISSUE TITLE
MonthYear web component: Accessibility review

### ISSUE THREAD

**Issue by k80bowman** [department-of-veterans-affairs/va.gov-team#39727]
## Description
We need to conduct a full accessibility review of the new Month Year variant of the Date web component. It can be viewed in Storybook here: https://design.va.gov/storybook/?path=/docs/components-va-date--default#month-year

## Acceptance Criteria
- [x] Full accessibility review of the Month Year date variant is completed
- [x] Add descriptions of any issues found to this ticket
    - [x] If any problems are found, add screen recordings and/or screenshots to comments on this ticket that show the problem.
- [x] Alert @k80bowman and @bkjohnson when the review is complete
- [ ] Re-review any fixes the developer makes to the component

**Comment by davidakennedy** [comment]
I've reviewed this component, and have a few points of feedback. As a reminder, here's the feedback framework we use when doing accessibility audits. It's handy here too.

### Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements

### Must

**Remove unneeded `aria-label` attributes**

All of the inputs have wrapper containers with `aria-label` on them, adding additional text for screen reader users. These are:

- Month: "aria-label="Please enter two digits for the month"
- Day: "aria-label="Please enter two digits for the day"
- Year: "aria-label="Please enter four digits for the year"

These cause issues with WCAG 4.1.2 Name, Role, Value. Labels should be associated with interactive elements with roles. Generic `div` and `span` elements aren't interactive and would need a role.

My recommendation here, for this demo, is to remove  the `aria-label` attributes, and probably not allow them there. It would be a better experience if text like that was available in the `legend` element for all users to see.

![Source code showing two unneeded aria-label attributes](https://user-images.githubusercontent.com/1473618/178072298-1d362a2f-8bd6-4679-9fae-332a501da8e9.png)

---

**Adjust duplicate ID attributes**

Both of the `select` elements have the same ID attributes. This might be an oversight, since this is a demo, but in production, the components should generate unique IDs. If multiple elements on the page have the same IDs, this could cause problem with assistive technology.

---

**Add aria-invalid when errors get triggered**

Adding `aria-invalid="true"` to each input that has an error would help assistive technology users better identify which input has the problem that needs to be corrected. Related side note: This might be a bigger effort if we don't do this already in components across the design system.

## Should

**Remove `aria-labelledby' attributes**

The select input has `aria-labelledby="select-label"` or similar. However, this isn't needed since each input is already associated with labels via the `for` attribute like this, `for="select"`. The `aria-labelledby` attributes should be removed.

![Red arrow pointing to source code of unneeded aria-labelledby attribute](https://user-images.githubusercontent.com/1473618/178072705-3e1cc568-53c8-4edc-bc34-7b4b06297711.png)

---

**Associate errors with the fields**

The error message here gets announced by `aria-live` as an alert when it occurs. However, that can get lost when/if a screen reader user uses forms mode in NVDA or Jaws to browse around form inputs only. Associating the error message with the input helps here, even though it's not perfect.

This can be done with `aria-describedby` on the input(s). Like: `<input aria-describedby="error-msg-id">`.

---

**Remove unneeded `aria-live` when errors occur**

One thing Tiffany discovered in testing this along with me was that there was an aria-live region to communicate if someone entered the wrong number of digits. Like: <small aria-live="polite" part="validation">(Min. 4 characters)</small> The problem is that it would never be announced, because it only occurs when an error is triggered. That means a role="alert" is already on the page. And that takes precedence in getting announced.

We should remove this additional `aria-live` region because multiple ARIA live regions on a page can get complex, and not all of them will be announced if they occur at the same time. Also, related to the above recommendation on contextual error messages, if this information is important to screen reader users, it may be important to everyone.

![Red arrow pointing to unneeded aria-live attribute](https://user-images.githubusercontent.com/1473618/178072978-f3672da0-7308-4a14-bfcf-871f735a8e86.png)

## Consider

**Add more contextual error messages**

- The error messages cover the entire date, not each individual input. This means when validating errors, you could fix the year, and still have it be red to signify the error. For example, the error message says: "Please enter a valid date" but it's really missing the year only. This may help improve the experience when it comes to fixing errors.

---

**How are multiple errors treated on a page?**

I don't know if any past research or discovery work has been done in the area, but it might be helpful as a long term strategy for error messages, to think through what happens when multiple inline errors occur. A well-thought out approach can help everyone, but in the accessibility realm, especially those with cognitive disabilities and those who use screen readers.

---

**Add Responsive styles**

I didn't see any responsive styles kick in for this component when I looked at it on a small screen. That may be a to do item or a non-issue, but wanted to note it. Not sure if it would come into play here, but it might.

If anything is unclear here, please let me know.

cc: @k80bowman @bkjohnson

**Comment by bkjohnson** [comment]
If the problem with the `aria-labels` is the missing role, would it be a good idea to add a `role` attribute to appropriate web components [like we do in `<va-radio>`](https://github.com/department-of-veterans-affairs/component-library/blob/75b1f96434274c65919fd6bcba76b0293263dc4b/packages/web-components/src/components/va-radio/va-radio.tsx#L175)? They are custom elements and aren't meant to be generic like `<div>` and `<span>`.

---

> Adjust duplicate ID attributes

This isn't an oversight, I [left a comment](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/844#issuecomment-1180586984) on another accessibility review talking about the situation a bit. I'm under the impression that if the IDs are in separate DOMs then it isn't an issue - is that wrong?

**Comment by bkjohnson** [comment]
Re: the `aria-labels`:

> Can we ask that question from a design standpoint first? Should this information be available to all users?

I don't know why they were added. I've assumed they were added to solve a problem, or requested by someone, but I haven't been able to find a concrete reason for them to be there. Ideally we would know the reason so that we can be sure that removing them doesn't break something for someone.

---

Thank you for the additional info on associating error messages with inputs, I have a better understanding now. Would you want something like this where we put an `aria-describedby` on each of the child components?

```diff
diff --git a/packages/web-components/src/components/va-date/va-date.tsx b/packages/web-components/src/components/va-date/va-date.tsx
index b5b03fa..2d5ef42 100644
--- a/packages/web-components/src/components/va-date/va-date.tsx
+++ b/packages/web-components/src/components/va-date/va-date.tsx
@@ -215,13 +215,14 @@ export class VaDate {
           </legend>
           <slot />
           {error && (
-            <span class="error-message" role="alert">
+            <span class="error-message" id="error-message" role="alert">
               <span class="sr-only">Error</span> {error}
             </span>
           )}
           <div class="date-container">
             <va-select
               label="Month"
+              aria-describedby={error ? 'error-message' : undefined}
               name={`${name}Month`}
               // Value must be a string
               value={month?.toString()}
@@ -238,6 +239,7 @@ export class VaDate {
             {!monthYearOnly && (
               <va-select
                 label="Day"
+                aria-describedby={error ? 'error-message' : undefined}
                 name={`${name}Day`}
                 // If day value set is greater than amount of days in the month
                 // set to empty string instead
@@ -256,6 +258,7 @@ export class VaDate {
             )}
             <va-text-input
               label="Year"
+              aria-describedby={error ? 'error-message' : undefined}
               name={`${name}Year`}
               maxlength={4}
               minlength={4}
```

As a reminder, we're using shadow DOM here, so any ID passed to an `aria-describedby` must be within the same DOM. We can't have the `<input>` inside of the `<va-text-input>` point to an id that exists outside of its scope, i.e. in the parent `<va-date>`.

**Comment by davidakennedy** [comment]
Re: aria-labels:

> I don't know why they were added. I've assumed they were added to solve a problem, or requested by someone, but I haven't been able to find a concrete reason for them to be there. Ideally we would know the reason so that we can be sure that removing them doesn't break something for someone.

Good call! I don't want to break anything either. I [raised this issue in Slack](https://dsva.slack.com/archives/C023ZKDJJK1/p1657747784483259) to see if anyone knows of any reason these are there.

> Would you want something like this where we put an aria-describedby on each of the child components?

Yes, that's perfect. And that makes sense, regarding the shadow DOM.

Thanks for the work here, @bkjohnson!

**Comment by bkjohnson** [comment]
Closed by https://github.com/department-of-veterans-affairs/component-library/pull/489, will open up a new ticket to track the "Consider" enhancements.
